### ISSUE TITLE
How to eliminate the second stage of training in RTMDet

### ISSUE THREAD

**Issue by AndreaPi** [open-mmlab/mmdetection#10544]
When training RTMDet, my loss goes way up at epoch 280, the mAP goes slightly down #10540 and I end up with a worse mAP and loss at the end of training (epoch 300). 

![image](https://github.com/open-mmlab/mmdetection/assets/8233615/6d931256-60f4-44c5-af06-5a9c76c6a3f9)
![image](https://github.com/open-mmlab/mmdetection/assets/8233615/f4284684-8d95-4721-97a3-4efbd9c142d8)

I made a lot of tests, and it seems to be due to this code in 
 https://github.com/open-mmlab/mmdetection/blob/main/configs/rtmdet/rtmdet_l_8xb32-300e_coco.py, which apparently switches the initial training pipeline for a new one at epoch 280:

```
train_pipeline_stage2 = [
    dict(type='LoadImageFromFile', backend_args={{_base_.backend_args}}),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(
        type='RandomResize',
        scale=(640, 640),
        ratio_range=(0.1, 2.0),
        keep_ratio=True),
    dict(type='RandomCrop', crop_size=(640, 640)),
    dict(type='YOLOXHSVRandomAug'),
    dict(type='RandomFlip', prob=0.5),
    dict(type='Pad', size=(640, 640), pad_val=dict(img=(114, 114, 114))),
    dict(type='PackDetInputs')
]
.
.
.
max_epochs = 300
stage2_num_epochs = 20
base_lr = 0.004
interval = 10
.
.
.
custom_hooks = [
    dict(
        type='EMAHook',
        ema_type='ExpMomentumEMA',
        momentum=0.0002,
        update_buffers=True,
        priority=49),
    dict(
        type='PipelineSwitchHook',
        switch_epoch=max_epochs - stage2_num_epochs,
        switch_pipeline=train_pipeline_stage2)
]
```

In order to prevent the pipeline switching, I added the following code to my custom config file:

```
# no stage 2 training
custom_hooks = [
    dict(
        type='PipelineSwitchHook',
        switch_epoch=n_epochs,
        switch_pipeline=train_pipeline)
    ]
```

1. Is this the correct way to do it?
2. Is it ok to prevent the pipeline switching? Looking at the mAP curve (see above) it looks like it only worsens performance, but I wonder if I'm losing some positive side-effect.

**Comment by Macok** [comment]
1. You can just delete `PipelineSwitchHook` entry from `custom_hooks` list.
2. I successfully trained models with 90%+ accuracy using RTMDet without pipeline switching. I believe it only makes sense on huge, complex datasets like COCO.

Consider I'm not mmdetection author though.

**Comment by AndreaPi** [comment]
Hi @Macok,

thanks for the reply! Can you show me how I should modify my config file in order to delete `PipelineSwitchHook`? Note that there's no `PipelineSwitchHook` in my config, but the switch still happens, likely because `PipelineSwitchHook` is defined in the base config.

```
_base_ = [
    '/mmdetection/configs/rtmdet/rtmdet_s_8xb32-300e_coco.py',
    ]

experiment_id = 'simple_rtmdet_s_8xb32-300e'


classes = ('A', 'B')
num_classes = len(classes)
metainfo = {
    'classes':
        classes,
    'palette':
        [(190, 77, 37), (37, 150, 190)]
}

model = dict(
    data_preprocessor=dict(
        type='DetDataPreprocessor',
        mean=[213.308, 102.466, 48.287],
        std=[31.949, 31.084, 20.839],
        bgr_to_rgb=False,
        batch_augments=None,
        pad_size_divisor=32),
    bbox_head=dict(num_classes=num_classes)
)

image_scale = (512, 672)
batch_size = 4
n_workers = 4
n_gpu = 1
base_lr = 0.004
eta = base_lr * (batch_size * n_gpu / 16)
n_epochs = 300

# pipelines setting
packed_inputs_items = ('img_id', 'img_path', 'img_shape', 'scale_factor', 'ori_shape')
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', scale=image_scale, keep_ratio=True),
    dict(type='PackDetInputs', meta_keys=packed_inputs_items),
]
val_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', scale=image_scale, keep_ratio=True),
    dict(type='PackDetInputs', meta_keys=packed_inputs_items),
]
test_pipeline = val_pipeline

# dataset settings
dataset_type = 'CocoDataset'
data_root = '/data/xxx/'
train_annotations = 'annotations/train.json'
val_annotations = 'annotations/val.json'
test_annotations = 'annotations/test.json'
image_data_prefix = data_root + 'images/'
backend_args = None

workers_configs = {
    "num_workers": n_workers,
    "persistent_workers": True,
}
dataset_configs = {
    "type": dataset_type,
    "metainfo": metainfo,
    "data_root": data_root,
    "data_prefix": dict(img=image_data_prefix),
    "backend_args": backend_args
}

# dataloader settings
train_dataloader = dict(
    batch_size=batch_size,
    sampler=dict(type='DefaultSampler', shuffle=True),
    **workers_configs,
    dataset=dict(
        ann_file=train_annotations,
        filter_cfg=dict(filter_empty_gt=True, min_size=32),
        pipeline=train_pipeline,
        **dataset_configs
    )
)
val_dataloader = dict(
    batch_size=1,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    **workers_configs,
    dataset=dict(
        ann_file=val_annotations,
        pipeline=val_pipeline,
        test_mode=True,
        **dataset_configs)
)
test_dataloader = dict(
    batch_size=1,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    **workers_configs,
    dataset=dict(
        ann_file=test_annotations,
        pipeline=test_pipeline,
        test_mode=True,
        **dataset_configs)
)

# evaluator settings
val_evaluator = dict(
    type='CocoMetric',
    ann_file=data_root + val_annotations,
    metric='bbox',
    format_only=False,
    backend_args=backend_args
)
test_evaluator = dict(
    type='CocoMetric',
    ann_file=data_root + test_annotations,
    metric='bbox',
    format_only=False,
    backend_args=backend_args
)

# optimizer settings
optimizer = dict(
    type='AdamW',
    lr=eta,
    weight_decay=0.0001
)
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=optimizer,
)

# workflow and workdir settings
train_cfg = dict(
    type='EpochBasedTrainLoop',
    max_epochs=n_epochs,
    val_interval=1)
val_cfg = dict(type='ValLoop')
test_cfg = dict(type='TestLoop')

# checkpoint: save the best model
default_hooks = dict(
    checkpoint=dict(
        type='CheckpointHook',
        interval=-1,
        save_best='auto'),
    logger=dict(
        type='LoggerHook',
        interval=50)
)

workflow = [('train', 1), ('val', 1)]

work_dir = f'logs/{experiment_id}'

# logger settings
vis_backends = [
    dict(type='LocalVisBackend'),
    dict(type='TensorboardVisBackend'),
]
visualizer = dict(
    type='DetLocalVisualizer',
    vis_backends=vis_backends,
    name='visualizer'
)
```

**Comment by GeorgePearse** [comment]
You probably want to override the custom_hooks. 

In this file (what you've pasted), include whatever you want and remove the PipelineSwitch. In your case, having looked at the rmdet_s file, it's just: 
```
custom_hooks = [
    dict(
        type='EMAHook',
        ema_type='ExpMomentumEMA',
        momentum=0.0002,
        update_buffers=True,
        priority=49),
]
```

**Comment by GeorgePearse** [comment]
I tend to init the config and then override values in the training script. 

e.g. 
```
cfg = Config.fromfile(config_file)
cfg.custom_hooks = [....]
```

Isolates my my changes from the original training template.

**Comment by AndreaPi** [comment]
> In this file (what you've pasted), include whatever you want and remove the PipelineSwitch. In your case, having looked at the rmdet_s file, it's just:

yep, that's what I just found out this morning. Thanks for your independent confirmation!

**Comment by AndreaPi** [comment]
> I tend to init the config and then override values in the training script.
> 
> e.g.
> 
> ```
> cfg = Config.fromfile(config_file)
> cfg.custom_hooks = [....]
> ```
> 
> Isolates my changes from the original training template.


I'm not sure I follow. Why should I define a `cfg` object, and where should I do it? I train my models from the CLI with

`python /mmdetection/tools/train.py ./configs/${MYCONFIG}.py`

where `${MYCONFIG}.py` is my custom config template which imports from the original training template thanks to the `base` assignment at the beginning, i.e.

```
_base_ = [
    '/mmdetection/configs/rtmdet/rtmdet_s_8xb32-300e_coco.py',
    ]
```

so I never modify the original training template. I guess you're doing something else? PS I'm closing the issue b/c my original problem was solved, but feel free to continue commenting with your answers

**Comment by GeorgePearse** [comment]
The cfg object is defined in 
```
/mmdetection/tools/train.py
```
That is what I'm referring to.
